From 6e81b8e39c950398e856bc18691abb01792f0150 Mon Sep 17 00:00:00 2001
From: Davor Joja <davorjoja@logicbricks.com>
Date: Mon, 23 Jun 2014 19:58:25 +0200
Subject: [PATCH 153/456] gpu: drm: xylon: Separated generic property
 parameters and functions

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

Property parameters, list and range creation functions placed to
separated source and header files.
Updated Makefile, fixed line align and objects ordering to keep
functionality grouping as through whole drm driver.

Signed-off-by: Davor Joja <davorjoja@logicbricks.com>
(cherry picked from commit db980b7e8e56e68c0352a016051e565634c4800d)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/gpu/drm/xylon/Makefile         |    6 +-
 drivers/gpu/drm/xylon/xylon_property.c |  108 ++++++++++++++++++++++++++++++++
 drivers/gpu/drm/xylon/xylon_property.h |   47 ++++++++++++++
 3 files changed, 158 insertions(+), 3 deletions(-)
 create mode 100644 drivers/gpu/drm/xylon/xylon_property.c
 create mode 100644 drivers/gpu/drm/xylon/xylon_property.h

diff --git a/drivers/gpu/drm/xylon/Makefile b/drivers/gpu/drm/xylon/Makefile
index d472ada..36bdaaa 100644
--- a/drivers/gpu/drm/xylon/Makefile
+++ b/drivers/gpu/drm/xylon/Makefile
@@ -2,8 +2,8 @@
 # Makefile for the drm device driver.  This driver provides support for the
 # Direct Rendering Infrastructure (DRI) in XFree86 4.1.0 and higher.
 
-xylon_drm-y := xylon_drv.o xylon_fbdev.o xylon_fb.o xylon_crtc.o \
-				xylon_plane.o xylon_logicvc.o xylon_encoder.o \
-				xylon_connector.o xylon_irq.o
+xylon_drm-y := xylon_drv.o xylon_crtc.o xylon_plane.o xylon_encoder.o \
+	       xylon_connector.o xylon_logicvc.o xylon_fbdev.o xylon_fb.o \
+	       xylon_irq.o xylon_property.o
 
 obj-$(CONFIG_DRM_XYLON) += xylon_drm.o
diff --git a/drivers/gpu/drm/xylon/xylon_property.c b/drivers/gpu/drm/xylon/xylon_property.c
new file mode 100644
index 0000000..3321ebe
--- /dev/null
+++ b/drivers/gpu/drm/xylon/xylon_property.c
@@ -0,0 +1,108 @@
+/*
+ * Xylon DRM property
+ *
+ * Copyright (C) 2014 Xylon d.o.o.
+ * Author: Davor Joja <davor.joja@logicbricks.com>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <drm/drm_crtc.h>
+
+#include "xylon_property.h"
+
+const struct drm_prop_enum_list property_layer_update[] = {
+	{ 0, "Layer Update Disable" },
+	{ 1, "Layer Update Enable" },
+	{/* end */}
+};
+
+const struct drm_prop_enum_list property_pixel_data_polarity[] = {
+	{ 0, "Pixel Data Polarity Normal" },
+	{ 1, "Pixel Data Polarity Invert" },
+	{/* end */}
+};
+
+const struct drm_prop_enum_list property_pixel_data_trigger[] = {
+	{ 0, "Pixel Data Trigger Falling" },
+	{ 1, "Pixel Data Trigger Rising" },
+	{/* end */}
+};
+
+const struct drm_prop_enum_list property_control[] = {
+	{ 0, "Plane Disable" },
+	{ 1, "Plane Enable" },
+	{/* end */}
+};
+
+const struct drm_prop_enum_list property_color_transparency[] = {
+	{ 0, "Plane Color Transparency Disable" },
+	{ 1, "Plane Color Transparency Enable" },
+	{/* end */}
+};
+
+const struct drm_prop_enum_list property_interlace[] = {
+	{ 0, "Plane Interlace Disable" },
+	{ 1, "Plane Interlace Enable" },
+	{/* end */}
+};
+
+const struct drm_prop_enum_list property_pixel_format[] = {
+	{ 0, "Plane ABGR Format Disable" },
+	{ 1, "Plane ABGR Format Enable" },
+	{/* end */}
+};
+
+int xylon_drm_property_size(const struct drm_prop_enum_list *list)
+{
+	int i = 0;
+
+	while (list[i].name != NULL)
+		i++;
+
+	return i;
+}
+
+int xylon_drm_property_create_list(struct drm_device *dev,
+				   struct drm_mode_object *obj,
+				   struct drm_property **prop,
+				   const struct drm_prop_enum_list *list,
+				   const char *name,
+				   int size)
+{
+	if (*prop)
+		return 0;
+
+	*prop = drm_property_create_enum(dev, 0, name, list, size);
+	if (*prop == NULL)
+		return -EINVAL;
+
+	drm_object_attach_property(obj, *prop, 0);
+
+	return 0;
+}
+
+int xylon_drm_property_create_range(struct drm_device *dev,
+				    struct drm_mode_object *obj,
+				    struct drm_property **prop,
+				    const char *name,
+				    u64 min, u64 max, u64 init)
+{
+	if (*prop)
+		return 0;
+
+	*prop = drm_property_create_range(dev, 0, name, min, max);
+	if (*prop == NULL)
+		return -EINVAL;
+
+	drm_object_attach_property(obj, *prop, init);
+
+	return 0;
+}
diff --git a/drivers/gpu/drm/xylon/xylon_property.h b/drivers/gpu/drm/xylon/xylon_property.h
new file mode 100644
index 0000000..e814f6e
--- /dev/null
+++ b/drivers/gpu/drm/xylon/xylon_property.h
@@ -0,0 +1,47 @@
+/*
+ * Xylon DRM property header
+ *
+ * Copyright (C) 2014 Xylon d.o.o.
+ * Author: Davor Joja <davor.joja@logicbricks.com>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef __XYLON_DRM_PROPERTY_H__
+#define __XYLON_DRM_PROPERTY_H__
+
+#define XYLON_DRM_PROPERTY_ALPHA_MIN 0
+#define XYLON_DRM_PROPERTY_ALPHA_MAX 255
+#define XYLON_DRM_PROPERTY_COLOR_MIN 0
+#define XYLON_DRM_PROPERTY_COLOR_MAX 0xFFFFFFFF
+
+extern const struct drm_prop_enum_list property_layer_update[];
+extern const struct drm_prop_enum_list property_pixel_data_polarity[];
+extern const struct drm_prop_enum_list property_pixel_data_trigger[];
+extern const struct drm_prop_enum_list property_control[];
+extern const struct drm_prop_enum_list property_color_transparency[];
+extern const struct drm_prop_enum_list property_interlace[];
+extern const struct drm_prop_enum_list property_pixel_format[];
+
+int xylon_drm_property_size(const struct drm_prop_enum_list *list);
+
+int xylon_drm_property_create_list(struct drm_device *dev,
+				   struct drm_mode_object *base,
+				   struct drm_property **prop,
+				   const struct drm_prop_enum_list *list,
+				   const char *name,
+				   int size);
+int xylon_drm_property_create_range(struct drm_device *dev,
+				    struct drm_mode_object *base,
+				    struct drm_property **prop,
+				    const char *name,
+				    u64 min, u64 max, u64 init);
+
+#endif /* __XYLON_DRM_PROPERTY_H__ */
-- 
1.7.5.4

